src/coord/src/catalog/storage.rs

```diff
@@ -438,6 +438,33 @@ impl Transaction<'_> {
         }
     }
 
+    fn assert_timestamp_bindings(&self, source_id: &GlobalId) {
+        let bindings: Vec<(PartitionId, Timestamp, MzOffset)> = self.inner
+            .prepare_cached(
+                "SELECT pid, timestamp, offset from timestamps where sid = ? order by pid, timestamp").unwrap()
+            .query_and_then(params![SqlVal(source_id)], |row| -> Result<_, Error> {
+                let partition: PartitionId = row.get::<_, String>(0)?.parse().expect("parsing partition id from string cannot fail");
+                let timestamp: Timestamp = row.get(1)?;
+                let offset = MzOffset {
+                    offset: row.get(2)?,
+                };
+
+                Ok((partition, timestamp, offset))
+            }).unwrap()
+            .collect::<Result<Vec<(PartitionId, Timestamp, MzOffset)>, _>>().unwrap();
+
+        let mut latest_offset = 0;
+        for (_pid, _ts, offset) in bindings.iter() {
```

Comment:
I think you need to reset `latest_offset` to zero when the partition changes -- like the following sequence of

(partition, time, offset) is correct iiuc but would fail the assert

(p1, t1, 100)
(p1, t2, 200)
(p2, t1, 50)